README.md
# WateringSystem


history.txt
2018-06-02 21:55:19.204170	WET


tests.py
import time
import datetime
import RPi.GPIO as GPIO

# settings
port_sensor_read = 18
port_sensor_cmd = 23
port_pump_cmd = 25

def initialize():
    GPIO.setmode(GPIO.BCM)
    GPIO.setwarnings(False)
    GPIO.setup(port_sensor_read, GPIO.IN)
    GPIO.setup(port_sensor_cmd, GPIO.OUT)
    GPIO.setup(port_pump_cmd, GPIO.OUT)


def checkIsWet():
    GPIO.output(port_sensor_cmd, GPIO.HIGH)
    time.sleep(1)
    value = GPIO.input(port_sensor_read) == 0
    GPIO.output(port_sensor_cmd, GPIO.LOW)
    return value


def doWork():
    initialize()
    print(checkIsWet())
    GPIO.cleanup()


def startPump():
    initialize()
    print('Pump ON')
    GPIO.output(port_pump_cmd, GPIO.HIGH)


def stopPump():
    initialize()
    GPIO.output(port_pump_cmd, GPIO.LOW)
    print('Pump OFF')
    GPIO.cleanup()
    



water.py
# External module imp
import RPi.GPIO as GPIO
import datetime
import time

init = False

GPIO.setmode(GPIO.BCM)

def getLastWatered():
    try:
        f = open("last_watered.txt", "r")
        return f.readline()
    except:
        return "NEVER"

def getStatus(pin = 8):
    GPIO.setup(pin, GPIO.IN)
    return GPIO.input(pin)

def initOutput(pin):
    GPIO.setup(pin, GPIO.OUT)
    GPIO.output(pin, GPIO.LOW)
    GPIO.output(pin, GPIO.HIGH)

def pumpOn(pin=7, delay=1):
    initOutput(pin)
    GPIO.output(pin, GPIO.LOW)
    time.sleep(delay)
    GPIO.output(pin, GPIO.HIGH)
    f = open("last_watered.txt", "w")
    f.write("Last watered {}".format(datetime.datetime.now()))
    f.close()


def autoWater(delay = 5, pumpPin = 7, sensorPin = 8):
    waterCount = 0
    initOutput(pumpPin)
    try:
        while waterCount < 10:
            time.sleep(delay)
            isWet = getStatus(sensorPin) == 0
            if isWet:
                waterCount = 0
            else:
                if waterCount < 5:
                    pumpOn(pumpPin, 1)
                waterCount+=1
    except KeyboardInterrupt:
        GPIO.cleanup()

status.py
import pickle
import os.path
import datetime
import sys

path = sys.path[0]
status_file = os.path.join(path, "status.bin")
print(sys.path)

class Status:
    auto = True
    feed_duration = 120
    feed_pause = 60
    feed_iterations = 2
    status = "PAUSE"
    last_watering = datetime.datetime.now()
    soil_was_wet_after = False
    last_reading = datetime.datetime.now()
    soil_was_wet = False

    def __init__(self):
        if os.path.isfile(status_file):
            self.load()
        else:
            self.auto = True
            self.feed_duration = 120
            self.feed_pause = 60
            self.feed_iterations = 2
            self.status = "PAUSE"
            self.last_watering = datetime.datetime.now()
            self.soil_was_wet_after = False
            self.last_reading = datetime.datetime.now()
            self.soil_was_wet = False
            self.save()


    def save(self):
        pickle.dump(self, open(status_file,'wb'))


    def load(self):
        value = pickle.load(open(status_file, 'rb'))
        print(value.feed_duration)
        self.auto = value.auto
        self.feed_duration = value.feed_duration
        self.feed_iterations = value.feed_iterations
        self.feed_pause = value.feed_pause
        self.last_reading = value.last_reading
        self.last_watering = value.last_watering
        self.soil_was_wet = value.soil_was_wet
        self.soil_was_wet_after = value.soil_was_wet_after
        self.status = value.status





webapp.py
from flask import Flask, flash, render_template, redirect, url_for, request, session, abort
import water_manager as mgr
import status
import os
import sys
import datetime



app = Flask(__name__)
path = sys.path[0]
app.secret_key = "a secret key"
status = status.Status()
log_filename = os.path.join(path, "log.txt")
history_filename = os.path.join(path, "history.txt")

def getModel(dateWatering='UNK', wasWet='UNK', isWetNow='UNK', statusMessage = "UNK"):
    status.load()
    viewModel = {
        'auto': "ON" if status.auto else "OFF",
        'dateWatering': status.last_watering.strftime("%x %X"),
        'wasWet': mgr.get_wet_dry(status.soil_was_wet_after),
        'isWet': mgr.get_wet_dry(mgr.checkIsWet()),
        'log': readLog(),
        'history': readHistory(),
        'events': [],
        'status': status.status,
        'title': 'WWWater @{0}'.format(datetime.datetime.now()),
        'day': (datetime.datetime.today()-datetime.timedelta(days=1)).day,
        'month':(datetime.datetime.today()-datetime.timedelta(days=1)).month,
        'year':(datetime.datetime.today()-datetime.timedelta(days=1)).year,
    }
    return viewModel


def readLog():
    result = '-'
    with open(log_filename) as f:
        line = f.readline()
        while line:
            result = '<p class='+getClass(line)+'>'+line.replace('\t', '&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;')+'</p>' + result
            line = f.readline()
    return result

def readHistory():
    result = '-'
    with open(history_filename) as f:
        line = f.readline()
        while line:
            result = '<p>'+line+'</p>'+result
            line = f.readline()
    return result

def getClass(line):
    result = "text-dark"
    if "INFO" in line:
        result = "text-dark"
    if "WARN" in line:
        result = "text-warning"
    if "ERROR" in line:
        result = "text-danger"
    return result

def checkLogin():
    b = 'logged_in' in session
    print(b)
    if not 'logged_in' in session:
        return render_template('login.html')
    else:
        return True

@app.route("/")
def index():
    if not 'logged_in' in session:
        return redirect(url_for('login'))
    else:
        templateDate = getModel()
        return render_template('main.html', **templateDate)


@app.route("/action/clear_log")
def clearLog():
    if checkLogin():
        open('log.txt', 'w')
        return redirect(url_for('index'))

@app.route("/action/clear_history")
def clearHistory():
    if checkLogin():
        open('history.txt', 'w')
        return redirect(url_for('index'))

@app.route("/action/start")
def startAutoWatering():
    if checkLogin():
        status.auto = True
        status.save()
        return redirect(url_for('index'))

@app.route("/action/stop")
def stopAutoWatering():
    if checkLogin():
        status.auto = False
        status.save()
        return redirect(url_for('index'))

@app.route("/action/stopNow")
def stopNow():
    if checkLogin():
        mgr.autoIsOn = False
        mgr.stopWork()
        return redirect(url_for('index'))

@app.route("/action/feed")
def feed():
    if checkLogin():
        mgr.feed()
        return redirect(url_for('index'))

@app.route("/action/read")
def readSensor():
    if checkLogin():
        return redirect(url_for('index'))


@app.route('/login', methods=['GET','POST'])
def login():
    if request.method == 'POST':
        if request.form['password'] == 'DICflDmpmp_1' and request.form['username'] == 'admin':
            session['logged_in'] = True
            return redirect(url_for('index'))
    else:
        return render_template('login.html')
if __name__ == "__main__":
    app.run(host='0.0.0.0', port=80, debug=True)

water_manager.py
import time
import datetime
import RPi.GPIO as GPIO
import status
import sys
import os
import tailhead

status = status.Status()
path = sys.path[0]
# settings
port_sensor_read = 18
port_sensor_cmd = 23
port_pump_cmd = 25

feed_duration = status.feed_duration
feed_pause = status.feed_pause
feed_iterations = status.feed_iterations

log_filename = os.path.join(path, "log.txt")
history_filename = os.path.join(path, "history.txt")
settings_filename = os.path.join(path, "settings.txt")
events_filename = os.path.join(path, "events.txt")

autoIsOn = status.auto


def initialize():
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(port_sensor_read, GPIO.IN)
    GPIO.setup(port_sensor_cmd, GPIO.OUT)
    GPIO.setup(port_pump_cmd, GPIO.OUT)


def checkIsWet():
    initialize()
    GPIO.output(port_sensor_cmd, GPIO.HIGH)
    time.sleep(2)
    value = GPIO.input(port_sensor_read) == 0
    GPIO.output(port_sensor_cmd, GPIO.LOW)
    writeHistory(value)
    status.last_reading = datetime.datetime.now()
    status.soil_was_wet = value
    status.save()
    GPIO.cleanup()
    return value


def turnOnPump():
    initialize()
    GPIO.output(port_pump_cmd, GPIO.HIGH)


def turnOffPump():
    initialize()
    GPIO.output(port_pump_cmd, GPIO.LOW)


def feed():
    writeLog(path, "INFO")
    turnOnPump()
    status.status = "FEEDING"
    status.save()
    writeLog("Watering for {0}s...".format(feed_duration), "INFO")
    time.sleep(feed_duration)
    turnOffPump()
    status.status = "PAUSE"
    status.last_watering = datetime.datetime.now()
    status.save()



def writeLog(message, errorLevel):
    line = str(datetime.datetime.now()) + "\t" + errorLevel + \
        "\t\t" + message + "\r\n"
    file = open(log_filename, 'a+')
    print(line)
    file.write(line)
    file.close()

def writeHistory(value):
    line = str(datetime.datetime.now())+"\t"+get_wet_dry(value)+"\r\n"
    file = open(history_filename, 'a+')
    file.write(line)
    file.close()


def get_wet_dry(isWet):
    return "WET" if isWet else "DRY"


def stopWork():
    GPIO.cleanup()
    autoIsOn = False
    status.auto = autoIsOn
    status.save()

def last_reads_are_dry(count):
    x = tailhead.tail(open(history_filename, 'rb'), count)
    dryCount = 0
    for i in x:
        if i.decode("utf-8").count("DRY")>0:
            dryCount += 1
    return dryCount == count

def shouldWork(isWet):
    result = False
    
    if not isWet:
        now = datetime.datetime.now()
        if (now.hour >= 19 and now.hour <=23) or (now.hour >= 3 and now.hour<=9):
            if last_reads_are_dry(5):
                result = True
    return result

def doWork():
    status.load()
    try:
        autoIsOn = status.auto
        isWet = checkIsWet()
        sw = shouldWork(isWet)
        s = "AUTO: {0}; SOIL: {1}; SHOULD_WORK: {2}".format(autoIsOn, get_wet_dry(isWet), sw)
        writeLog(s, "INFO" if isWet else "WARN")
        if sw:
            writeLog("Start watering.", "INFO")
            feed()
            isWet = checkIsWet()
            status.last_watering = datetime.datetime.now()
            status.soil_was_wet_after = isWet
            status.save()
            writeLog("After watering the soil is {0}.".format(get_wet_dry(isWet)), "INFO" if isWet else "ERROR")
    except KeyboardInterrupt:
        writeLog("Cleanup GPIO ports", "INFO")
        GPIO.cleanup()
        status.status = "SLEEPING"
        status.save()

if __name__ == '__main__':
    writeLog("Manager called from crontab", "INFO")
    doWork()

wtf.py
import time
import datetime
import RPi.GPIO as GPIO
import status

status = status.Status()

# settings
port_sensor_read = 18
port_sensor_cmd = 23
port_pump_cmd = 25

feed_duration = status.feed_duration
feed_pause = status.feed_pause
feed_iterations = status.feed_iterations

log_filename = "log.txt"
settings_filename = "settings.txt"
events_filename = "events.txt"

shouldWork = status.auto


def initialize():
    print("init")
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(port_sensor_read, GPIO.IN)
    GPIO.setup(port_sensor_cmd, GPIO.OUT)
    GPIO.setup(port_pump_cmd, GPIO.OUT)


def checkIsWet():
    initialize()
    writeLog("Sensor ON", "INFO")
    GPIO.output(port_sensor_cmd, GPIO.HIGH)
    time.sleep(1)
    writeLog("Read Sensor", "INFO")
    value = GPIO.input(port_sensor_read) == 0
    writeLog("Sensor OFF", "INFO")
    GPIO.output(port_sensor_cmd, GPIO.LOW)
    writeLog("Value read: {0}".format(value), "INFO")
    status.last_reading = datetime.datetime.now()
    status.soil_was_wet = value
    return value


def turnOnPump():
    initialize()
    writeLog("Pump ON", "INFO")
    GPIO.output(port_pump_cmd, GPIO.HIGH)


def turnOffPump():
    initialize()
    writeLog("Pump OFF", "INFO")
    GPIO.output(port_pump_cmd, GPIO.LOW)


def feed():
    writeLog("Turn ON pump", "INFO")
    turnOnPump()
    status.status = "FEEDING"
    status.save()
    writeLog("Feeding for {0}s...".format(feed_duration), "INFO")
    time.sleep(feed_duration)
    writeLog("Turn OFF pump", "INFO")
    turnOffPump()
    status.status = "PAUSE"
    status.save()
    writeLog("Pause for {0}s...".format(feed_pause), "INFO")
    time.sleep(feed_pause)


def writeLog(message, errorLevel):
    line = str(datetime.datetime.now()) + "\t" + errorLevel + \
        "\t\t" + message + "\r\n"
    file = open(log_filename, 'a+')
    print(line)
    file.write(line)
    file.close()


def stopWork():
    GPIO.cleanup()
    shouldWork = False

def doWork():
    status.load()
    writeLog("START", "INFO")
    try:
        shouldWork = status.auto
        if shouldWork:
            counter = 1
            isWet = checkIsWet()
            writeLog("Soil is wet? {0}".format(isWet), "INFO" if isWet else "WARN")
            while not isWet and counter <= feed_iterations:
                writeLog("Feeding START #{0}/{1}".format(counter, feed_iterations), "INFO")
                feed()
                status.last_watering = datetime.datetime.now()
                status.save()
                writeLog("Feeding END #{0}".format(counter), "INFO")
                isWet = checkIsWet()
                status.soil_was_wet_after = isWet
                status.save()

                writeLog("Soil is still dry? {0}".format(not isWet), "INFO" if isWet else "WARN")
                counter += 1
            if not isWet:
                writeLog("After {0} feeding cycles the soil is still dry. Check the water level in tank!".format(feed_iterations), "ERROR")
        else:
            writeLog("AUTO is off", "INFO")
    except KeyboardInterrupt:
        writeLog("Cleanup GPIO ports", "INFO")
        GPIO.cleanup()
        status.status = "SLEEPING"
        status.save()

if __name__ == '__main__':
    writeLog("STARTING MANAGER...", "INFO")
    doWork()
